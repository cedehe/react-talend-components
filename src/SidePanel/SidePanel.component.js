import React from 'react';
import classNames from 'classnames';
import { Button } from 'react-bootstrap';
import theme from './SidePanel.scss';
import Icon from '../Icon';

/* eslint-disable jsx-a11y/href-no-hash */

/**
 * This component aims to display links as a menu.
 * @param {object} props react props
 *
 @example
 const actions = [
	 { label: 'Preparations', icon: 'fa fa-asterisk', onClick: action('Preparations clicked') },
	 { label: 'Datasets', icon: 'fa fa-file-excel-o', onClick: action('Datasets clicked') },
	 { label: 'Favorites', icon: 'fa fa-star', onClick: action('Favorites clicked') }
 ];
 <SidePanel
	 actions={ actions }
	 onToggleDock={ action('Toggle dock clicked') }
	 docked={ isDocked }
     toggleIcon={ 'fa fa-arrow-left' }
 />
 *
 */
function SidePanel(props) {
	const actions = props.actions || [];

	const dockedCSS = { [theme.docked]: props.docked };
	const navCSS = classNames(
		theme['tc-side-panel'],
		dockedCSS,
		'tc-side-panel'
	);
	const listCSS = classNames(
		'nav nav-pills nav-inverse nav-stacked',
		'tc-side-panel-list',
		theme['action-list']
	);

	return (
		<nav className={navCSS}>
			<Button
				id={props.id && `${props.id}-toggle-dock`}
				className={theme['toggle-btn']}
				bsStyle="link"
				onClick={props.onToggleDock}
				aria-hidden="true"
				title="Toggle side panel"
			>
				<Icon name={props.toggleIcon || 'talend-arrow-left'} />
			</Button>
			<ul className={listCSS}>
				{actions.map(action => (
					<li
						key={action.label}
						className={classNames(
							'tc-side-panel-list-item',
							{ active: !!action.active }
							)}
					>
						<Button
							id={props.id && `${props.id}-nav-${action.label.toLowerCase().split(' ').join('-')}`}
							title={`Display ${action.label}`}
							bsStyle="link"
							role="link"
							className={theme.link}
							onClick={action.onClick}
						>
							{ action.icon ? <Icon name={action.icon} /> : null }
							{ props.docked ? null : <span>{action.label}</span> }
						</Button>
					</li>
				))}
			</ul>
		</nav>
	);
}

SidePanel.propTypes = {
	id: React.PropTypes.string,
	actions: React.PropTypes.arrayOf(
		React.PropTypes.shape({
			label: React.PropTypes.string,
			icon: React.PropTypes.string,
			onClick: React.PropTypes.func,
		})
	),
	onToggleDock: React.PropTypes.func,
	docked: React.PropTypes.bool,
	toggleIcon: React.PropTypes.string,
};

export default SidePanel;
