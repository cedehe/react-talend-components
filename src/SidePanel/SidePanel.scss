////
/// Side panel
/// @group Custom widgets
////

/// Side panel toggle button color
/// @type Color
$tc-side-panel-toggle-color: #DDD !default;

/// Side panel default spacing
/// @type Number (Unit)
$tc-side-panel-spacing: 20px !default;

/// Docked side panel spacing
/// @type Number (Unit)
$tc-side-panel-docked-spacing: 15px !default;

.tc-side-panel {
	overflow-x: hidden;
	padding: 0 $tc-side-panel-spacing;
	transition: 0.2s padding ease-out;

	.action-list {
		margin: 0 (-$tc-side-panel-spacing);
		transition: 0.2s margin ease-out;
	}

	.toggle-btn {
		position: relative;
		display: block;
		padding: 0;
		margin: $tc-side-panel-spacing 0;
		line-height: 0.5;
		color: $tc-side-panel-toggle-color;
		box-shadow: -2px 0 0 $tc-side-panel-toggle-color;

		& :global(.tc-svg-icon) {
			fill: $tc-side-panel-toggle-color;
		}

		&:hover{
			background-color: inherit;
		},
		&:focus,
		&:active {
			color: inherit;
			outline: 0;
		}

		i,
		:global(.tc-svg-icon) {
			transition: 0.2s transform ease-out;
		}

		// Increase click area
		:after {
			position: absolute;
			content: '';
			top: 0;
			right: 0;
			bottom: 0;
			left: 0;
			margin: -1rem;
			border-radius: 50%;
		}
	}

	.link {
		text-align: left;
		font-size: 14px;
		text-transform: uppercase;

		i + span,
		:global(.tc-svg-icon) + span {
			padding-left: 20px;
		}

		:global(.tc-svg-icon) {
			fill: currentColor;
		}

		span {
			transition: 0.2s opacity linear;
		}
	}

	&.docked {
		padding: 0 $tc-side-panel-docked-spacing;

		.toggle-btn i,
		.toggle-btn :global(.tc-svg-icon) {
			transform: rotate(-180deg);
		}

		.link span {
			opacity: 0;
		}
	}
}
