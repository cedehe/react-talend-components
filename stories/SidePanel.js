import React from 'react';
import { storiesOf, action } from '@kadira/storybook';

import { SidePanel, IconsProvider } from '../src/index';

const actions = [
	{
		label: 'Preparations',
		icon: 'fa fa-asterisk',
		onClick: action('Preparations clicked'),
		active: true,
	},
	{
		label: 'Datasets',
		icon: 'fa fa-file-excel-o',
		onClick: action('Datasets clicked'),
	},
	{
		label: 'Favorites',
		icon: 'fa fa-star',
		onClick: action('Favorites clicked'),
	},
];

storiesOf('Side Panel', module)
	.addWithInfo('default', () => {
		return (
			<div>
				<h1>SidePanel</h1>
				<h2>Definition</h2>
				<a href="http://guidelines.talend.com/document/92132#/ui-controls/side-panel">Spec</a>
				<h2>Examples</h2>
				<IconsProvider />
				<SidePanel
					actions={actions}
					onToggleDock={action('Toggle dock clicked')}
					docked={false}
				/>
			</div>
		);
	})
	.addWithInfo('docked', () =>
		(
			<div>
				<h1>SidePanel</h1>
				<h2>Definition</h2>
				<a href="http://guidelines.talend.com/document/92132#/ui-controls/side-panel">Spec</a>
				<h2>Examples</h2>
				<IconsProvider />
				<SidePanel
					actions={actions}
					onToggleDock={action('Toggle dock clicked')}
					docked
				/>
			</div>
		)
	)
	.addWithInfo('custom toggle icon', () =>
		(
			<div>
				<h1>SidePanel</h1>
				<h2>Definition</h2>
				<a href="http://guidelines.talend.com/document/92132#/ui-controls/side-panel">Spec</a>
				<h2>Examples</h2>
				<IconsProvider />
				<SidePanel
					actions={actions}
					onToggleDock={action('Toggle dock clicked')}
					toggleIcon="fa-arrow-left"
					docked
				/>
			</div>
		)
	);
