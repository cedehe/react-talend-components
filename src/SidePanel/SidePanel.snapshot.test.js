import React from 'react';
import renderer from 'react-test-renderer';
import SidePanel from './SidePanel.component';

jest.mock('react-dom');

describe('SidePanel', () => {
	it('should render expanded menu', () => {
		// given
		const actions = [
			{ label: 'Preparations', icon: 'fa fa-asterisk' },
			{ label: 'Datasets', icon: 'fa fa-file-excel-o' },
			{ label: 'Favorites', icon: 'fa fa-star', active: true },
		];
		const docked = false;

		// when
		const sidePanel = <SidePanel actions={actions} docked={docked} />;
		const wrapper = renderer.create(sidePanel).toJSON();

		// then
		expect(wrapper).toMatchSnapshot();
	});

	it('should render docked menu', () => {
		// given
		const actions = [
			{ label: 'Preparations', icon: 'fa fa-asterisk' },
			{ label: 'Datasets', icon: 'fa fa-file-excel-o' },
			{ label: 'Favorites', icon: 'fa fa-star', active: true },
		];
		const docked = true;

		// when
		const sidePanel = <SidePanel actions={actions} docked={docked} />;
		const wrapper = renderer.create(sidePanel).toJSON();

		// then
		expect(wrapper).toMatchSnapshot();
	});
});
