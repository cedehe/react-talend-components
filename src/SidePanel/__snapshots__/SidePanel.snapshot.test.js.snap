exports[`SidePanel should render docked menu 1`] = `
<nav
  className="undefined tc-side-panel">
  <button
    aria-hidden="true"
    className="btn btn-link"
    disabled={false}
    id={undefined}
    onClick={undefined}
    title="Toggle side panel"
    type="button">
    <i
      aria-hidden="true"
      className="fa fa-arrow-left"
      title={null} />
  </button>
  <ul
    className="nav nav-pills nav-inverse nav-stacked tc-side-panel-list">
    <li
      className="tc-side-panel-list-item">
      <button
        className="btn btn-link"
        disabled={false}
        id={undefined}
        onClick={undefined}
        role="link"
        title="Display Preparations"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-asterisk"
          title={null} />
      </button>
    </li>
    <li
      className="tc-side-panel-list-item">
      <button
        className="btn btn-link"
        disabled={false}
        id={undefined}
        onClick={undefined}
        role="link"
        title="Display Datasets"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-file-excel-o"
          title={null} />
      </button>
    </li>
    <li
      className="tc-side-panel-list-item active">
      <button
        className="btn btn-link"
        disabled={false}
        id={undefined}
        onClick={undefined}
        role="link"
        title="Display Favorites"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-star"
          title={null} />
      </button>
    </li>
  </ul>
</nav>
`;

exports[`SidePanel should render expanded menu 1`] = `
<nav
  className="tc-side-panel">
  <button
    aria-hidden="true"
    className="btn btn-link"
    disabled={false}
    id={undefined}
    onClick={undefined}
    title="Toggle side panel"
    type="button">
    <i
      aria-hidden="true"
      className="fa fa-arrow-left"
      title={null} />
  </button>
  <ul
    className="nav nav-pills nav-inverse nav-stacked tc-side-panel-list">
    <li
      className="tc-side-panel-list-item">
      <button
        className="btn btn-link"
        disabled={false}
        id={undefined}
        onClick={undefined}
        role="link"
        title="Display Preparations"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-asterisk"
          title={null} />
        <span>
          Preparations
        </span>
      </button>
    </li>
    <li
      className="tc-side-panel-list-item">
      <button
        className="btn btn-link"
        disabled={false}
        id={undefined}
        onClick={undefined}
        role="link"
        title="Display Datasets"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-file-excel-o"
          title={null} />
        <span>
          Datasets
        </span>
      </button>
    </li>
    <li
      className="tc-side-panel-list-item active">
      <button
        className="btn btn-link"
        disabled={false}
        id={undefined}
        onClick={undefined}
        role="link"
        title="Display Favorites"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-star"
          title={null} />
        <span>
          Favorites
        </span>
      </button>
    </li>
  </ul>
</nav>
`;

exports[`SidePanel should render id when props.id provided 1`] = `
<nav
  className="undefined tc-side-panel">
  <button
    aria-hidden="true"
    className="btn btn-link"
    disabled={false}
    id="context-toggle-dock"
    onClick={undefined}
    title="Toggle side panel"
    type="button">
    <i
      aria-hidden="true"
      className="fa fa-arrow-left"
      title={null} />
  </button>
  <ul
    className="nav nav-pills nav-inverse nav-stacked tc-side-panel-list">
    <li
      className="tc-side-panel-list-item">
      <button
        className="btn btn-link"
        disabled={false}
        id="context-nav-preparations"
        onClick={undefined}
        role="link"
        title="Display Preparations"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-asterisk"
          title={null} />
      </button>
    </li>
    <li
      className="tc-side-panel-list-item">
      <button
        className="btn btn-link"
        disabled={false}
        id="context-nav-datasets"
        onClick={undefined}
        role="link"
        title="Display Datasets"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-file-excel-o"
          title={null} />
      </button>
    </li>
    <li
      className="tc-side-panel-list-item active">
      <button
        className="btn btn-link"
        disabled={false}
        id="context-nav-favorites"
        onClick={undefined}
        role="link"
        title="Display Favorites"
        type="button">
        <i
          aria-hidden="true"
          className="fa fa-star"
          title={null} />
      </button>
    </li>
  </ul>
</nav>
`;
